go/chat/attachment_crypt.go

```diff
@@ -10,36 +16,83 @@ type Encrypter interface {
 	// Encrypt takes a plaintext reader and returns a ciphertext reader.
 	Encrypt(plaintext io.Reader) (ciphertext io.Reader)
 
-	// Key returns the ephemeral key that was used during Encrypt.
-	Key() []byte
+	// EncryptKey returns the ephemeral key that was used during Encrypt.
+	EncryptKey() []byte
+
+	// VerifyKey returns the public portion of the signing key that
+	// can be used for signature verification.
+	VerifyKey() []byte
 }
 
 type Decrypter interface {
-	// Decrypt takes a ciphertext reader and a key and returns a
-	// plaintext reader.
-	Decrypt(ciphertext io.Reader, key []byte) (plaintext io.Reader, err error)
+	// Decrypt takes a ciphertext reader, encryption and verify keys.
+	// It returns a plaintext reader.
+	Decrypt(ciphertext io.Reader, encKey, verifyKey []byte) (plaintext io.Reader)
 }
 
-// PassThrough implements Encrypter and Decrypter.  It does no encryption
-// or decryption.
-type PassThrough struct{}
+var nonce = &[signencrypt.NonceSize]byte{
+	0x10, 0x10, 0x10, 0x10, 0x20, 0x20, 0x20, 0x20,
+	0x30, 0x30, 0x30, 0x30, 0x40, 0x40, 0x40, 0x40,
+}
 
-func NewPassThrough() *PassThrough {
-	return &PassThrough{}
+type SignEncrypter struct {
+	encKey    signencrypt.SecretboxKey
+	signKey   signencrypt.SignKey
+	verifyKey signencrypt.VerifyKey
 }
 
-func (p *PassThrough) EncryptedLen(size int) int {
-	return size
+func NewSignEncrypter() (*SignEncrypter, error) {
+	enc, err := libkb.GenerateNaclDHKeyPair()
```

Comment:
This isn't quite the right function to use. This is generating an encryption public-private key _pair_ rather than a simple symmetric encryption key. What we want is 32 random bytes, probably from `crypto/rand`. `GenerateNaclDHKeyPair` will pretty much give us that, but with a couple downsides:

- Depending on the implementation, it might clamp a few of the bits to zero. This has to do with some of the safety guarantees of Curve25519, which I don't really understand :p
- It will do an unnecessary curve multiplication to compute the public key, which is kind of expensive.